chore (doc):Create a files for "Building a generic plugin" and link "Files gatsby looks for in plugin"  closes #20527 (#21961)
gatsbyjs/gatsby: docs/docs/files-gatsby-looks-for-in-a-plugin.md

---

> title: Creating a Generic plugin

Reviewer: I personally think this will confuse people who aren't creating generic plugins, as it will make it harder to find this information. Is it possible to create two pages to serve different (yet overlapping) use cases?

---

This section is aimed at explaining the structure of a gatsby plugin and the files that are needed to create it.

As seen in the [what is a plugin doc](/docs/what-is-a-plugin/) is a piece of software that acts as an add-on and gives your site additional functionality.

They contain a file, usually in the project root, called `package.json` - this file holds various metadata relevant to the project. This file is used to give information to npm that allows it to identify the project as well as handle the project's dependencies.

> The process of initializing a `package.json` in a project is done by running :

Reviewer 2: ```suggestion

The process of initializing a `package.json` in a project is done by running:
```

```shell
  npm init
```

## What files does Gatsby look for in a plugin?

> After running the initialization command. You would need to provide information for the `package.json`

Reviewer 2: Since `npm init` lets you add in information for a bunch of these fields maybe this should say something like this:

```suggestion
You can provide more information for fields in the `package.json` that weren't prompted by running the initialization command.
```

- `gatsby-browser.js` — usage details are in the [browser API reference](/docs/browser-apis/)
- `gatsby-node.js` — usage details are in the [Node API reference](/docs/node-apis/)
- `gatsby-ssr.js` — usage details are in the [SSR API reference](/docs/ssr-apis/)

In a generic plugin the `gatsby-node.js` file enables the use of [gatsby node APIs](/docs/node-apis/) which is where all the processes like createPage,createResolvers,sourceNodes exist.

> In this file you are able to carry out functions that load API keys, send calls to APIs,create Gatsby-nodes using the API response, Create pages to display responses and loop through responses to display an individual response. (A good use case of this would be a plugin that gets data from an API.)

Reviewer 2: ```suggestion
In this file you are able to carry out functions that load API keys, send calls to APIs,  create Gatsby-nodes using the API response, create pages to display responses, and loop through responses to display an individual response (a good use case of this would be a plugin that gets data from an API).
```
